cmd/gapit/status.go

```diff
@@ -0,0 +1,292 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"runtime"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/event/task"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/service"
+)
+
+type statusVerb struct {
+	StatusFlags
+}
+
+func init() {
+	verb := &statusVerb{}
+	verb.StatusUpdateInterval = 1000
+
+	app.AddVerb(&app.Verb{
+		Name:      "status",
+		ShortHelp: "Attaches to an existing gapis, and provides status",
+		Action:    verb,
+	})
+}
+
+func clear() {
+	if runtime.GOOS == "windows" {
+		cmd := exec.Command("cmd", "/c", "cls")
+		cmd.Stdout = os.Stdout
+		cmd.Run()
+	} else {
+		cmd := exec.Command("clear")
+		cmd.Stdout = os.Stdout
+		cmd.Run()
+	}
+}
+
+func (verb *statusVerb) Run(ctx context.Context, flags flag.FlagSet) error {
+	client, err := getGapis(ctx, verb.Gapis, GapirFlags{})
+	if err != nil {
+		return log.Err(ctx, err, "Failed to connect to the GAPIS server")
+	}
+	defer func() {
+		if err := client.Close(); err != nil {
+			log.E(ctx, "Error closing client: %v", err)
+		}
+	}()
+
+	statusMutex := sync.Mutex{}
+
+	type tsk struct {
+		id         uint64
+		parentID   uint64
+		name       string
+		background bool
+		progress   int32
+		blocked    bool
+		children   map[uint64]*tsk
+	}
+
+	ancestors := make(map[uint64][]uint64)
+	activeTasks := make(map[uint64]*tsk)
+	totalBlocked := 0
+
+	var findTask func(*map[uint64]*tsk, []uint64) *tsk
+
+	findTask = func(base *map[uint64]*tsk, indices []uint64) *tsk {
```

Comment:
Probably no need to make it a pointer to map?

Reply:
Fair, I constantly get confused by golang semantics for maps, as they are non-obvious.

Reply:
Fixed